### PR TITLE
SSR: Use document-head to populate `title`, `meta`s, and `link`s; drop `Helmet`

### LINKED ISSUE #3795: Themes: Get rid of `Helmet`
We should really get rid of Helmet, as it's more hassle than benefit.

Rationale: For most pages we want to give `title` and `meta` tags, we'll have to fetch some data from Redux (e.g. theme name for a theme sheet). From there, we have to add those data to Helmet; `rewind` Helmet right after `renderToString` to get the data back (and not forget to cache them alongwith the rendered markup!); and pass them to `index.jade` to have them rendered in their appropriate places.

Or we just get them from Redux, and pass them to `index.jade`.

### PR DESCRIPTION
**Pull request by ockham**
This PR:

* has the theme sheet and showcase use `DocumentHead` instead of `Head` (which was a rather specialized component tailored to the theme showcase's SEO needs, based on `react-helmet``
* drops `Head`, and the `react-helmet` dependency
* modifies `server/render` to use data from the `document-head` state to populate SEO relevant fields (`title`, `meta`, `link`) for SSR
* removes `description` related reducer and actions.
 * Description meta can be handled by just using `addMeta`, so I don't think it's worthwhile having specialized functions around.

To test:
* While logged out, check a given theme sheet's page source. Verify that SEO relevant tags are present (`title`, `meta`, `link`). (Compare with `master` or `production`.)
* While logged out, navigate to the showcase (http://calypso.localhost:3000/design), and click on any theme. Watch the title change -- at no point should it not make any sense.
* Same when logged in. (Also, there obviously shouldn't be any errors, console or otherwise.)

TODO (separate PR):
* Make sure we SSR `title` et al for `/design` (_this isn't currently done on `master`, so not a regression_)

Fixes #3795

Test live: https://calypso.live/?branch=remove/head

### CODE REVIEW

File: client/my-sites/theme/main.jsx [hunk 1]
```diff
@@ -348,46 +348,58 @@ const ThemeSheet = React.createClass( {
 		const analyticsPageTitle = `Themes > Details Sheet${ section ? ' > ' + titlecase( section ) : '' }${ siteID ? ' > Site' : '' }`;
 
 		const { name: themeName, description } = this.props;
-		const title = i18n.translate( '%(themeName)s Theme', {
+		const title = themeName && i18n.translate( '%(themeName)s Theme', {
```
Comment (aduth) [inline client/my-sites/theme/main.jsx]: What was the reason for the `themeName &&` change? This won't prevent the `<DocumentHead />` from dispatching the title, will it?

Per: https://github.com/Automattic/wp-calypso/blob/9923003f10d4e89f120bab7d9bce2bc49dcb42fa/client/components/data/document-head/index.jsx#L29-L31

File: client/my-sites/themes/theme-showcase.jsx [hunk 2]
```diff
@@ -25,16 +45,18 @@ const optionShape = PropTypes.shape( {
 
 const ThemeShowcase = React.createClass( {
 	propTypes: {
+		tier: PropTypes.oneOf( [ '', 'free', 'premium' ] ).isRequired,
```
Comment (aduth) [inline client/my-sites/themes/theme-showcase.jsx]: Is `isRequired` redundant / conflicting with `getDefaultProps`?

File: client/state/document-head/reducer.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ import {
 export function title( state = '', action ) {
 	switch ( action.type ) {
 		case DOCUMENT_HEAD_TITLE_SET:
-			return action.title;
+			return action.title || state;
```
Comment (aduth) [inline client/state/document-head/reducer.js]: Ah, I guess this is why it's okay to dispatch with `false`? But as mentioned in a previous pull request, there are cases where we want to intentionally reset the title to `undefined` (#5897). In any case, I'd rather we avoid dispatching altogether if we don't intend for the title to be used.

**Comment by aduth** [comment]
Looking forward to this 🎉

File: client/my-sites/theme/main.jsx (hunk 1 above)
Comment (ockham) [inline client/my-sites/theme/main.jsx]: > What was the reason for the `themeName &&` change?

Basically to retain logic that was previously [here](https://github.com/Automattic/wp-calypso/pull/7600/files#diff-f576b3c13daa17f9ccee514c4c46bc07L359).

File: client/state/document-head/reducer.js (hunk 3 above)
Comment (ockham) [inline client/state/document-head/reducer.js]: > I'd rather we avoid dispatching altogether if we don't intend for the title to be used.

Yeah. The issue here is that returning `undefined` from a reducer yields an error:

> Reducer "title" returned undefined handling "DOCUMENT_HEAD_TITLE_SET"

Comment (ockham) [inline client/state/document-head/reducer.js]: I guess this might mean I need to guard agains `undefined` title in `<DocumentHead/>`'s `componentWillReceiveProps()` after all ([related](https://github.com/Automattic/wp-calypso/pull/7224#discussion_r74437843)).

File: client/my-sites/themes/theme-showcase.jsx (hunk 2 above)
Comment (ockham) [inline client/my-sites/themes/theme-showcase.jsx]: Yep, I'll remove it.

File: client/state/document-head/reducer.js (hunk 3 above)
Comment (ockham) [inline client/state/document-head/reducer.js]: @aduth: Fix in 960f9cc8c6ff56a60dc91807eff160b0b65416e6.

A condition like `if ( 'title' in this.props ) { ... }` (as originally suggested [here](https://github.com/Automattic/wp-calypso/pull/7224#r74437843)) doesn't catch `this.props.title` being undefined, which would hence be used when dispatching the action.

Comment (ockham) [inline client/state/document-head/reducer.js]: Hrm, I forgot about this part

> there are cases where we want to intentionally reset the title to `undefined` (#5897)

Anyway, we just can't dispatch `title` being set to `undefined`, or we'll get [that error](https://github.com/Automattic/wp-calypso/pull/7600#r76866662).

Comment (aduth) [inline client/state/document-head/reducer.js]: What about a separate action creator and action type specifically for unsetting the title?

Comment (ockham) [inline client/state/document-head/reducer.js]: Sounds good, I'll give that a try.

File: client/my-sites/themes/theme-showcase.jsx [hunk 4]
```diff
@@ -82,16 +104,24 @@ const ThemeShowcase = React.createClass( {
...
+		const metas = [
+			{ name: 'description', property: 'og:description', content: themesMeta[ tier ].description },
+			{ property: 'og:url', content: themesMeta[ tier ].canonicalUrl },
+			{ property: 'og:type', content: 'website' }
+		];
+
+		// FIXME: Logged-in title should only be 'Themes'
 		return (
 			<Main className="themes">
+				<DocumentHead title={ themesMeta[ tier ].title } metas={ metas } />
```
Comment (aduth) [inline client/my-sites/themes/theme-showcase.jsx]: The prop for metas is `meta`, not `metas`.

File: client/state/document-head/selectors.js [hunk 5]
```diff
@@ -78,3 +80,25 @@ export function getFormattedTitle( state ) {
...
 }
+
+/**
+ * Returns an array of document meta objects as set by the DocumentHead
+ * component or addDocumentHeadMeta action.
+ *
+ * @param  {Object}  state  Global state tree
+ * @return {Object[]}       Array of meta objects
+ */
+export function getMeta( state ) {
```
Comment (aduth) [inline client/state/document-head/selectors.js]: Hmm, in the context of global selectors, should they come to reality, would `getMeta` be sufficiently unique? Seems less so than `getTitle`, though a similar argument could be made there.

File: server/render/index.js [hunk 6]
```diff
@@ -78,13 +70,19 @@ export function serverRender( req, res ) {
...
+	if ( config.isEnabled( 'server-side-rendering' ) && ! context.user ) {
+		if ( context.layout ) {
+			const key = context.renderCacheKey || JSON.stringify( context.layout );
+			Object.assign( context, render( context.layout, key ) );
+		}
+		if ( context.store ) {
+			const title = getFormattedTitle( context.store.getState() );
+			const metas = getMeta( context.store.getState() );
+			const links = getLink( context.store.getState() );
+			context.head = { title, metas, links };
```
Comment (aduth) [inline server/render/index.js]: We wrap this in `context.store` condition, but the Jade template assumes `head.title` will always exist. Do we need to be safer there too? Or remove the `context.store` check?

Comment (ockham) [inline server/render/index.js]: We can't just remove the `context.store` check or `context.store.getState()` might fail. Fixing this with f7ffb39 by falling back to `undefined` attrs.

**Comment by ockham** [comment]
Okay, addressed feedback, including renaming of all selectors. Some have pretty long names now...

File: client/state/document-head/selectors.js [hunk 7]
```diff
@@ -19,7 +21,7 @@ const UNREAD_COUNT_CAP = 40;
  * @param  {Object}  state  Global state tree
  * @return {?String}        Document title
  */
-export function getTitle( state ) {
+export function getDocumentHeadTitle( state ) {
```
Comment (aduth) [inline client/state/document-head/selectors.js]: `client/reader/full-post/controller.js` still imports as `getTitle` in this branch.

File: server/render/index.js [hunk 8]
```diff
@@ -78,13 +70,23 @@ export function serverRender( req, res ) {
...
+
+		if ( context.store ) {
+			title = getDocumentHeadFormattedTitle( context.store.getState() );
+			metas = getDocumentHeadMeta( context.store.getState() );
+			links = getDocumentHeadLink( context.store.getState() );
+
+			context.initialReduxState = pick( context.store.getState(), 'documentHead', 'ui', 'themes' );
+		}
+
+		context.head = { title, metas, links };
```
Comment (aduth) [inline server/render/index.js]: Do the defaults and `context.head` assignment need to be outside of the `config` / `context.user` condition block too to ensure it's always set?

Comment (ockham) [inline server/render/index.js]: Good call. This might've gone unnoticed in testing since the feature flag is on in all environments, and  `context.user` is notoriously unset on the server in `development`.

**Comment by ockham** [comment]
Addressed feedback and rebased (for node 3.7.0).

File: server/pages/index.jade [hunk 9]
```diff
@@ -11,10 +11,8 @@ doctype html
 
 html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width' : '')
 	head
-		if helmetTitle
```
Comment (aduth) [inline server/pages/index.jade]: This condition is still necessary. If I navigate to e.g. the editor locally in this branch, there's an empty `<title></title>` in the `<head>`. Either that or we add the default to Redux and pull it from there in `server/render`.

title

Comment (ockham) [inline server/pages/index.jade]: > Either that or we add the default to Redux and pull it from there in `server/render`.

I'll opt for the latter since this is consistent with what I'm already doing with the `og:site_name` `meta`.

Comment (ockham) [inline server/pages/index.jade]: Done. Note that this means an `initialReduxState` will now be created and passed in the skeleton HTML also for non-isomorphic sites, but it's fairly small for non-iso sections.

File: server/render/index.js [hunk 10]
```diff
@@ -73,20 +70,34 @@ export function render( element, key = JSON.stringify( element ) ) {
...
+		title = getDocumentHeadFormattedTitle( context.store.getState() );
+		metas = getDocumentHeadMeta( context.store.getState() );
+		links = getDocumentHeadLink( context.store.getState() );
+
+		let reduxSubtrees = [ 'documentHead' ];
+		if ( isSectionIsomorphic( context.store.getState() ) ) {
+			reduxSubtrees = reduxSubtrees.concat( [ 'ui', 'themes' ] );
+		}
+
+		context.initialReduxState = pick( context.store.getState(), ...reduxSubtrees );
```
Comment (aduth) [inline server/render/index.js]: Not a big deal, but `...reduxSubtrees` doesn't need to be spread. Lodash will accept an array of keys to pick as the second argument.

File: server/render/index.js [hunk 11]
```diff
@@ -73,20 +65,29 @@ export function render( element, key = JSON.stringify( element ) ) {
...
 
-	if ( config.isEnabled( 'server-side-rendering' ) &&
-		context.store &&
-		context.layout &&
-		! context.user ) {
-		context.initialReduxState = pick( context.store.getState(), 'ui', 'themes' );
-		const key = context.renderCacheKey || JSON.stringify( context.layout );
-		Object.assign( context, render( context.layout, key ) );
+	if ( config.isEnabled( 'server-side-rendering' ) && ! context.user ) {
+		if ( context.layout ) {
```
Comment (seear) [inline server/render/index.js]: Interesting that the `context.layout` and `context.store` branches have been split apart here, because I don't thing we ever want to do one without the other?

**Comment by seear** [comment]
This is looking and testing fine from a themes/SSR point of view.

File: server/pages/index.jade (hunk 9 above)
Comment (seear) [inline server/pages/index.jade]: > Done. Note that this means an initialReduxState will now be created and passed in the skeleton HTML also for non-isomorphic sites, but it's fairly small for non-iso sections.

Whatever is in `initialReduxState` is not picked up from the local state cache, so we need to be careful. This will just be `ui` and `documentHead`?

File: server/render/index.js (hunk 11 above)
Comment (ockham) [inline server/render/index.js]: We do now, see the [discussion](https://github.com/Automattic/wp-calypso/pull/7600#discussion_r81335085).

tl;dr: With the new logic that populates the SSR'd `title` and `meta`s, we want to use the `document-head` subtree's data also for non-iso sections.

File: server/pages/index.jade (hunk 9 above)
Comment (ockham) [inline server/pages/index.jade]: For non-iso, it's only `documentHead`. For iso, it's unchanged (i.e. `documentHead`, `ui`, and `themes`).
